woocommerce/woocommerce
# [Experimental] Product Filters Chips style and new interactivity API implementation

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Closes #49975, closes #50868

This PR:

- Reimplements Interactivity API for filter blocks to be simpler and more predictable:
    - The filter URL is now manipulated only by the `woocommerce/product-filters` block; the URL is updated (if needed) whenever the context changes.
    - Filter blocks no longer update the URL directly, but update `woocommerce/product-filters` context.
    - The action exposed from the Attribute Filter block store (and soon other filter blocks) is element agnostic, which doesn't require the element of filter option to be an HTML input/checkbox element.

- Adds Chips block, which displays filter options as chips:
    - HTML button is used for chip for keyboard navigation.
    - `aria-checked` is used to distinguish select states.

- Uses a new styling approach for Chips and Checkbox List blocks:
    - The generated style only contains custom CSS variables.
    - In `style.scss`, use those variables for styling with a fallback to the current color.
    - Classes and styles are saved to block save markup.
    - Use HTML API to get the calculated style and class in the server-side render callback and include them in the returned content.

#### The flow of the new implementation

```mermaid
sequenceDiagram
    participant User
    participant UI
    participant Store
    participant URL

    User->>UI: Interacts with filter
    UI->>Store: Update filter parameters
    Store->>Store: Check if params changed
    Store->>URL: Update URL if params changed
    URL-->>User: Reflect new filter state
```

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.

2. Reset the Product Filters template part if needed.
3. Edit the Product Filters template part, see the Active and Attribute filter block.
4. Add another attribute filter block and set the style to Chips.
5. Save and test on the front end, see the client navigation works as expected when interacting with Active and (two) Attribute filter blocks.
6. Edit the template part again, and try setting custom colors for both Chips and List.
7. See the editor preview reflect custom color settings.
8. Save and test on the front end, see the custom color settings applied on the frontend too.




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
